src/main/java/com/android/volley/toolbox/BaseHttpStack.java

```diff
@@ -0,0 +1,95 @@
+
+import com.android.volley.AuthFailureError;
+import com.android.volley.Request;
+
+import org.apache.http.Header;
+import org.apache.http.ProtocolVersion;
+import org.apache.http.StatusLine;
+import org.apache.http.entity.BasicHttpEntity;
+import org.apache.http.message.BasicHeader;
+import org.apache.http.message.BasicHttpResponse;
+import org.apache.http.message.BasicStatusLine;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.SocketTimeoutException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/** An HTTP stack abstraction. */
+@SuppressWarnings("deprecation") // for HttpStack
+public abstract class BaseHttpStack implements HttpStack {
+
+    /**
+     * Performs an HTTP request with the given parameters.
+     *
+     * <p>A GET request is sent if request.getPostBody() == null. A POST request is sent otherwise,
+     * and the Content-Type header is set to request.getPostBodyContentType().
+     *
+     * @param request the request to perform
+     * @param additionalHeaders additional headers to be sent together with
+     *         {@link Request#getHeaders()}
+     * @return the {@link HttpResponse}
+     * @throws SocketTimeoutException if the request times out
+     * @throws IOException if another I/O error occurs during the request
+     * @throws AuthFailureError if an authentication failure occurs during the request
+     */
+    public abstract HttpResponse executeRequest(
+            Request<?> request, Map<String, String> additionalHeaders)
+            throws IOException, AuthFailureError;
+
+    /**
+     * @deprecated use {@link #executeRequest} instead to avoid a dependency on the deprecated
+     * Apache HTTP library. Nothing in Volley's own source calls this method. However, since
+     * {@link BasicNetwork#mHttpStack} is exposed to subclasses, we provide this implementation in
+     * case legacy client apps are dependent on that field. This method may be removed in a future
+     * release of Volley.
+     */
+    @Deprecated
+    @Override
+    public final org.apache.http.HttpResponse performRequest(
+            Request<?> request, Map<String, String> additionalHeaders)
+            throws IOException, AuthFailureError {
+        HttpResponse response = executeRequest(request, additionalHeaders);
+
+        ProtocolVersion protocolVersion = new ProtocolVersion("HTTP", 1, 1);
+        StatusLine statusLine = new BasicStatusLine(
+                protocolVersion, response.getStatusCode(), "" /* reasonPhrase */);
+        BasicHttpResponse apacheResponse = new BasicHttpResponse(statusLine);
+
+        List<Header> headers = new ArrayList<>();
+        for (Map.Entry<String, List<String>> entry : response.getHeaders().entrySet()) {
+            for (String value : entry.getValue()) {
+                headers.add(new BasicHeader(entry.getKey(), value));
+            }
+        }
+        apacheResponse.setHeaders(headers.toArray(new Header[0]));
```

Comment:
If you allocate the correct size up front, headers.toArray(new Header[headers.size()]), then you can avoid an unnecessary allocation of an array.

Reply:
Done.